src/main/java/frc/robot/subsystems/localization/Localizer.java

```diff
@@ -0,0 +1,193 @@
+package frc.robot.subsystems.localization;
+
+import edu.wpi.first.apriltag.*;
+import edu.wpi.first.cameraserver.CameraServer;
+import edu.wpi.first.cscore.CvSink;
+import edu.wpi.first.cscore.CvSource;
+import edu.wpi.first.cscore.UsbCamera;
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.math.geometry.Transform3d;
+import edu.wpi.first.wpilibj.Filesystem;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.robot.Constants;
+import org.photonvision.PhotonCamera;
+import org.photonvision.targeting.PhotonPipelineResult;
+import org.photonvision.targeting.PhotonTrackedTarget;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Optional;
+
+import org.opencv.core.Mat;
+import org.opencv.core.Point;
+import org.opencv.core.Scalar;
+import org.opencv.imgproc.Imgproc;
+
+public class Localizer {
+  private Thread visionThread;
+  private AprilTagDetector detector;
+  private AprilTagFieldLayout fieldLayout;
+  private NodePositionLayout nodeLayout;
+  private String filename = "/ChargedUp.json";
+  private String nodePositionFilename = "/ScoringLocations.json";
+  private AprilTagDetection[] detections;
+  private AprilTagPoseEstimator estimator;
+  int count;
+
+  public Localizer() throws IOException {
+    detector = new AprilTagDetector();
+    detector.addFamily("tag16h5", 0);
+    AprilTagDetector.Config config = new AprilTagDetector.Config();
+    detector.setConfig(config);
+    String path = Filesystem.getDeployDirectory().getPath() + filename;
+    fieldLayout = new AprilTagFieldLayout(path);
+
+    String nodePositionPath = Filesystem.getDeployDirectory().getPath() + nodePositionFilename;
+    nodeLayout = new NodePositionLayout(nodePositionPath);
+
+    AprilTagPoseEstimator.Config poseEstConfig =
+    new AprilTagPoseEstimator.Config(
+        Constants.TARGET_SIZE_METERS,
+        Constants.FX_PIXELS,
+        Constants.FY_PIXELS,
+        Constants.CX_PIXELS,
+        Constants.CY_PIXELS);
+estimator = new AprilTagPoseEstimator(poseEstConfig);
+
+
+PhotonCamera camera = new PhotonCamera("Slotheye");
+
+
+
+    visionThread =
+        new Thread(
+            () -> {
```

Comment:
Nothing in this thread really needs to be running in parallel. You're using to run a loop with update logic, but for this, wpilib already has `my_subsystem.periodic()`. Please move the contents of the thread into `periodic()`

As a side note, using threads come with extra risks, like when two different threads are reading/writing to the same variable at the same time. To avoid these risks use threads a little as possible. WPILib is designed so that threads should not be necessary in most cases.

Reply:
How should I move the contents into a periodic